python/repeated.c

```diff
@@ -0,0 +1,705 @@
+                                                   PyObject* args) {
+  return PyUpb_RepeatedContainer_Extend(_self, args);
+}
+
+// -----------------------------------------------------------------------------
+// RepeatedCompositeContainer
+// -----------------------------------------------------------------------------
+
+static PyObject* PyUpb_RepeatedCompositeContainer_AppendNew(PyObject* _self) {
+  PyUpb_RepeatedContainer* self = (PyUpb_RepeatedContainer*)_self;
+  upb_array* arr = PyUpb_RepeatedContainer_AssureWritable(self);
+  if (!arr) return NULL;
+  const upb_fielddef* f = PyUpb_RepeatedContainer_GetField(self);
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  const upb_msgdef* m = upb_fielddef_msgsubdef(f);
+  upb_msg* msg = upb_msg_new(m, arena);
+  upb_msgval msgval = {.msg_val = msg};
+  upb_array_append(arr, msgval, arena);
+  return PyUpb_CMessage_Get(msg, m, self->arena);
+}
+
+PyObject* PyUpb_RepeatedCompositeContainer_Add(PyObject* _self, PyObject* args,
+                                               PyObject* kwargs) {
+  PyUpb_RepeatedContainer* self = (PyUpb_RepeatedContainer*)_self;
+  PyObject* py_msg = PyUpb_RepeatedCompositeContainer_AppendNew(_self);
+  if (!py_msg) return NULL;
+  if (PyUpb_CMessage_InitAttributes(py_msg, args, kwargs) < 0) {
+    Py_DECREF(py_msg);
+    upb_array_delete(self->ptr.arr, upb_array_size(self->ptr.arr) - 1, 1);
+    return NULL;
+  }
+  return py_msg;
+}
+
+static PyObject* PyUpb_RepeatedCompositeContainer_Append(PyObject* _self,
+                                                         PyObject* value) {
+  if (!PyUpb_CMessage_Check(value)) return NULL;
+  PyObject* py_msg = PyUpb_RepeatedCompositeContainer_AppendNew(_self);
+  if (!py_msg) return NULL;
+  PyObject* none = PyUpb_CMessage_MergeFrom(py_msg, value);
+  if (!none) {
+    Py_DECREF(py_msg);
+    return NULL;
+  }
+  Py_DECREF(none);
+  return py_msg;
+}
+
+int AssignSubscript(PyObject* _self, PyObject* index, PyObject* value) {
+  PyUpb_RepeatedContainer* self = (PyUpb_RepeatedContainer*)_self;
+  size_t size =
+      PyUpb_RepeatedContainer_IsUnset(self) ? 0 : upb_array_size(self->ptr.arr);
+
+  if (value != NULL) {
+    PyErr_SetString(PyExc_TypeError, "does not support assignment");
+    return -1;
+  }
+
+  Py_ssize_t i, count, step;
+  if (!IndexToRange(index, size, &i, &count, &step)) return -1;
+  upb_array_delete(self->ptr.arr, i, count);  // XXX: step
+  return 0;
+}
+
+static PyObject* PyUpb_RepeatedContainer_Insert(PyObject* _self,
+                                                PyObject* args) {
+  PyUpb_RepeatedContainer* self = (PyUpb_RepeatedContainer*)_self;
+  Py_ssize_t index;
+  PyObject* value;
+  if (!PyArg_ParseTuple(args, "nO", &index, &value)) return NULL;
+  upb_array* arr = PyUpb_RepeatedContainer_AssureWritable(self);
+  if (!arr) return NULL;
+
+  // Normalize index.
+  Py_ssize_t size = upb_array_size(arr);
+  if (index < 0) index += size;
+  if (index < 0) index = 0;
```

Comment:
is this really the semantics of python?

Reply:
It does appear to be the semantics of the insert operation:

```
$ python
Python 3.9.8 (main, Nov  7 2021, 15:47:09) 
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x = list(range(10))
>>> x.insert(-25, 111)
>>> x
[111, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> x = list(range(10))
>>> x.insert(-5, 111)
>>> x
[0, 1, 2, 3, 4, 111, 5, 6, 7, 8, 9]
>>> x = list(range(10))
>>> x.insert(125, 111)
>>> x
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 111]
```